Flex_attention explicit GQA support

joydddd opened the pull request:
### tl;dr
This PR adds GQA support to higher order op `flex_attention`. 

## Details
When `enable_gqa` is set to True, HOP `flex_attention(score_mod, query, key, value, block_mask, enable_gqa)` runs Group Query Attention(GQA), where the number of query heads (Hq) is a multiple of number of key/value heads (Hkv). Each group of query heads (`Hq//Hkv` heads) attends to a shared kv head. 
Otherwise, `flex_attention` assumes Multi Head Attention (MHA) where the number of query heads is equal the number of kv heads. 

The `score_mod` API is adapted accordingly to take `q_head` as head index. 
```
def score_mod(score: torch.Tensor, batch: torch.Tensor, q_head: torch.Tensor, token_q: torch.Tensor, token_kv: torch.Tensor) -> torch.Tensor
```

## Example
```python
import torch
from torch.nn.attention._flex_attention import _flex_attention as flex_attention
from torch.nn.attention._flex_attention import _create_block_mask, _causal

torch.manual_seed(0)


def query_key_value_clones(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    dtype: torch.dtype = None,
):
    """Clones the query, key, and value tensors and moves them to the specified dtype."""
    if dtype is None:
        dtype = query.dtype
    query_ref = query.clone().detach().to(dtype).requires_grad_(query.requires_grad)
    key_ref = key.clone().detach().to(dtype).requires_grad_(key.requires_grad)
    value_ref = value.clone().detach().to(dtype).requires_grad_(value.requires_grad)
    return query_ref, key_ref, value_ref


# Lets create some input tensors
# The input tensor has shape (batch_size, num_heads, seq_len, head_dim).
# query and key/value can have different num_heads and seq_len
# Here 8 query heads share one KV head. 
query = torch.randn(2, 8, 2048, 64, device="cuda", dtype=torch.float32, requires_grad=True)
key = torch.randn(2, 2, 2048, 64, device="cuda", dtype=torch.float32, requires_grad=True)
value = torch.randn(2, 2, 2048, 64, device="cuda", dtype=torch.float32, requires_grad=True)

query1, key1, value1 = query_key_value_clones(query, key, value)

# Lets create a score_modification. We take alibi_bias as an example. 
# score_mod takes batch index, query head index, query index, and key/value index. 
def _generate_alibi_bias(num_kv_heads: int, num_q_heads: int):
    def _alibi_bias(
        score: torch.Tensor,
        b: torch.Tensor,
        hq: torch.Tensor,
        token_q: torch.Tensor,
        token_kv: torch.Tensor,
    ) -> torch.Tensor:
        # Let's calculate kv head from query head index
        group = num_q_heads // num_kv_heads
        hkv = hq // group

        scale = torch.exp2(-((hkv + 1) * 8.0 / num_kv_heads))
        return score + (token_kv - token_q) * scale

    return _alibi_bias


# Let's apply a casual mask on top of it
def casual_alibi_bias(score, batch, q_head, token_q, token_kv):
    score = _generate_alibi_bias(2, 8)(score, batch, q_head, token_q, token_kv)
    return torch.where(token_q <= token_kv, score, float("-inf"))


# Generate a block sparsity mask for our new score_mod function. 
# The mask is broadcasted long head & batch dimensions. 
block_mask = _create_block_mask(casual_alibi_bias, B=None, Hq=None, M=2048, N=2048)

# Lets call flex_attention with this new score modification
output = flex_attention(query, key, value, score_mod=casual_alibi_bias, block_mask=block_mask, enable_gqa=True)

compiled_flex_attention = torch.compile(flex_attention)
out_compiled = compiled_flex_attention(query1, key1, value1, score_mod=casual_alibi_bias, block_mask=block_mask, enable_gqa=True)

torch.testing.assert_close(output, out_compiled, atol=5e-2, rtol=2e-2)


```




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Chillee commented on torch/nn/attention/_flex_attention.py:
```diff
@@ -231,65 +231,72 @@ def _create_block_mask_from_mask(
     score_mod: _score_mod_signature,
     B: int,
-    H: int,
+    Hkv: int,
```
Why make it another dimension?

joydddd replied:
score_mod takes (b, hq, m, n) but the block mask is of shape (B, Hkv, (Hq//Hkv)*M, N) to matches our internal shape. We need the group size information to properly cast the hkv to hq. Maybe we should take a tuple of `(Hkv, Hq)` instead?

Chillee replied:
Can we not hide this from the user?

joydddd replied:
Hide this detail from user: broadcast mask along G dimension instead `_flex_attention` via `_broadcast_block_mask_for_gqa`.

Chillee commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -228,25 +228,35 @@ def build_subgraph_buffer(
-    SPARSE_Z = {{size("SPARSE_KV_NUM_BLKS", 0)}}
-    SPARSE_H = {{size("SPARSE_KV_NUM_BLKS", 1)}}
+    use_sparse_mask: tl.constexpr = (not SPARSE_Q_BLOCK_SIZE == 0) and (not SPARSE_KV_BLOCK_SIZE == 0)
+    if use_sparse_mask:
```
Out of curiosity, any particular reason you added this?

joydddd replied:
This allows an empty mask to have SPARSE_BLOCK_SIZE of 0 instead of seq_len. This allows seq_len to be dynamic, while SPARSE_BLOCK_SIZE remain static.

joydddd replied:
Reverted specific path for no mask. Fallback a full mask of BLOCK 128x128 when no block_mask is specified.

Chillee commented on third_party/kineto:
should remove this change.

joydddd replied:
Fixed.

joydddd commented on torch/nn/attention/_flex_attention.py:
```diff
@@ -304,15 +311,13 @@ def _create_block_mask(
         q_indices=torch.zeros([1, 1, 1, 1], dtype=torch.int32, device=device),
-        KV_BLOCK_SIZE=kv_len,
-        Q_BLOCK_SIZE=q_len,
+        KV_BLOCK_SIZE=0,
+        Q_BLOCK_SIZE=0,
```
Change the empty block mask to have KV_BLOCK_SIZE and Q_BLOCK_SIZE equals to 0 instead of seq_len. We want KV_BLOCK_SIZE & Q_BLOCK_SIZE to be statics, but seq_len can be dynamic.

joydddd commented on torch/nn/attention/_flex_attention.py:
```diff
@@ -371,19 +377,39 @@ def score_mod(
+        # mark head_dim & num of heads always to be static
         for x in [query, key, value]:
             torch._dynamo.mark_static(x, -1)
+            torch._dynamo.mark_static(x, -3)
```
Mark number of heads as statics. Only seq_len and batch_size can be dynamic.

joydddd commented:
Broadcast inside flex_attention kernel instead

joydddd committed:
Add explicit GQA support. TODO: debug cuda memory error

Attempt to add GQA to mask generation

Fix backward

Fix submodule

Reduce GQA unit test mem usage

Remove special empty mask pass in flex_attention kernel

Generate GQA block mask from input

move mask group braodcast to kernel

Fix Group dim reduction

Flex DQ

Fix backward DK DV

Revert changes in flex_attention

default is_gqa in test

joydddd committed:
Fix tests

joydddd committed:
Fix lint

Chillee commented on torch/nn/attention/_flex_attention.py:
```diff
@@ -359,8 +365,8 @@ def _create_block_mask_inner(
     score_mod: _score_mod_signature,
-    B: int,
-    H: int,
+    B: Optional[int],
```
Why is the function signature changed here?

Chillee commented on torch/nn/attention/_flex_attention.py:
```diff
@@ -472,12 +496,30 @@ def score_mod(
 
+    if not query.size(-1) == key.size(-1):
+        raise ValueError(
+            "NYI: Embedding dimension of the query and key must be the same"
```
This doesn't need to be a NYI - this is just always a requirement.

joydddd replied:
Resolved.

Chillee commented on torch/_higher_order_ops/flex_attention.py:
```diff
@@ -125,31 +125,45 @@ def math_attention(
     score_mod = torch.vmap(score_mod, in_dims=(0, None, None, None, 0) + in_dim_buffers)
     score_mod = torch.vmap(score_mod, in_dims=(0, None, None, 0, None) + in_dim_buffers)
     score_mod = torch.vmap(score_mod, in_dims=(0, None, 0, None, None) + in_dim_buffers)
+    score_mod = torch.vmap(score_mod, in_dims=(0, None, 0, None, None) + in_dim_buffers)
```
I think you might have mentioned, but why does this need to be done? Why can't the head index just represent the query head?

joydddd replied:
The query head dimension is reshaped into (Hkv, G) to allow K, V broadcast along G dimension. From `score_mod` 's perspective, head index does represent query head. That why both Hkv & G dimension are mapped to hq arg of `score_mod`.

Chillee reviewed (changes requested):
It's late/I'm sleepy, so sorry if we already went over this. But are the semantics for `score_mod(score, b, h, q_idx, kv_idx)` that `h` represents the query head? It seems that there's a bunch of places in this code that don't align with that.

joydddd commented:
> It's late/I'm sleepy, so sorry if we already went over this. But are the semantics for `score_mod(score, b, h, q_idx, kv_idx)` that `h` represents the query head? It seems that there's a bunch of places in this code that don't align with that.

This is my intension. Let me check if I missed anything.

joydddd committed:
rename API is_gqa->enable_gqa

joydddd commented:
Performance impact is minimal. The same as pre-reshape inputs: query=query.reshape((B, Hkv, -1, D))

joydddd committed:
Update score_mod to support GQA inputs

joydddd committed:
Fix benchmark Lint

joydddd committed:
interprate GQA instead Triton kernel

joydddd committed:
Fix dq

joydddd committed:
Fix backward grid

joydddd committed:
Fix typo

Chillee commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -613,20 +628,20 @@ def flex_attention(*args, **kwargs):
 
     return (
-        triton.cdiv(num_queries, meta["BLOCK_M2"])
+        triton.cdiv(num_queries, meta["BLOCK_M2"]) * (q_heads // kv_heads)
```
How does this generally work if `num_queries` is not a multiple of `BLOCK_M2`?

joydddd replied:
Well it doesn't work. We're throwing NYI on seq_len not a multiple of 128. And, this is confusing but `num_queries` refers to q_seq_len here. 😢

Chillee commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -613,20 +628,20 @@ def flex_attention(*args, **kwargs):
     return (
-        triton.cdiv(num_queries, meta["BLOCK_M2"])
+        triton.cdiv(num_queries, meta["BLOCK_M2"]) * (q_heads // kv_heads)
         + triton.cdiv(num_key_value, meta["BLOCK_N1"]),
```
Also, it seems that we're generally launching *less* blocks here? It might be cleaner to just split it up over the `kv`heads as well, and then launch a separate clean-up reduction.

Chillee reviewed (changes requested):
Mostly LGTM. I'd like to see whether folding the reduction inside the backwards kernel is worth it vs. just launching more blocks.

Since folding the reduction inside the backwards kernel also adds more code complexity.

joydddd commented:
Landed in https://github.com/pytorch/pytorch/pull/131559
